### PR TITLE
Preference 'xml.symbols.exclude' for an array of patterns to exclude symbols for file(s)

### PR DESCRIPTION
**Pull request by NikolasKomonen**
No tests yet

The structure of this PR.

So for the patterns it will take in an array of patterns as strings.
The server will convert those patterns to XMLExcludedSymbolFiles classes.

Each TextDocument has a doSymbols Boolean.
When symbols is requested on it it will check if that Boolean has been set yet, or else it will compute a pattern match.

Let me know if this works or could be improved.

### CODE REVIEW

**Comment by NikolasKomonen** [comment]
@angelozerr Are you fine with this implementation. I put a quick explanation at the top. I will do tests if you think this is good.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/settings/XMLFileAssociation.java [hunk 1]
```diff
@@ -49,32 +48,11 @@ public void setSystemId(String systemId) {
 	}
 
 	public boolean matches(String uri) {
-		try {
-			return matches(new URI(uri));
-		} catch (Exception e) {
-			return false;
-		}
+		return URIUtils.matches(pattern, uri, pathMatcher);
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/settings/XMLFileAssociation.java]: In your case pathMatcher is every time null. In the original XMLFileAssociation, the pathMatcher is null the first call of matches and initialized and the second time of call of matches it is initialized.

I think we should not have URIUtils.matches, but a base class for path matcher (like PatchMatcher class). XMLFileAssociation could extend PatchMatcher  and your XMLExcludedSymbolFile too

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import java.nio.file.PathMatcher;
+import java.util.Objects;
+
+import org.eclipse.lsp4xml.commons.TextDocument;
+import org.eclipse.lsp4xml.utils.URIUtils;
+
+/**
+ * XMLExcludedSymbolFiles
+ */
+public class XMLExcludedSymbolFile {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: It should extends the new class PathPatternMatcher

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+	public void setPattern(String pattern) {
+		this.pattern = pattern;
+	}
+
+	public boolean matches(String uri) {
+		return URIUtils.matches(pattern, uri, pathMatcher);
+	}
+
+	public boolean matches(URI uri) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+	public void setPathMatcher(PathMatcher pathMatcher) {
+		this.pathMatcher = pathMatcher;
+	}
+
+	public String getPattern() {
+		return pattern;
+	}
+
+	public void setPattern(String pattern) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+		this.pattern = pattern;
+	}
+
+	public boolean matches(String uri) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	
+	public PathMatcher getPathMatcher() {
+		return pathMatcher;
+	}
+
+	public void setPathMatcher(PathMatcher pathMatcher) {
+		this.pathMatcher = pathMatcher;
+	}
+
+	public String getPattern() {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	public XMLExcludedSymbolFile(String pattern) {
+		setPattern(pattern);
+	}
+
+	
+	public PathMatcher getPathMatcher() {
+		return pathMatcher;
+	}
+
+	public void setPathMatcher(PathMatcher pathMatcher) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+import org.eclipse.lsp4xml.utils.URIUtils;
+
+/**
+ * XMLExcludedSymbolFiles
+ */
+public class XMLExcludedSymbolFile {
+
+	private transient PathMatcher pathMatcher;
+
+	private String pattern;
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+import org.eclipse.lsp4xml.commons.TextDocument;
+import org.eclipse.lsp4xml.utils.URIUtils;
+
+/**
+ * XMLExcludedSymbolFiles
+ */
+public class XMLExcludedSymbolFile {
+
+	private transient PathMatcher pathMatcher;
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/URIUtils.java [hunk 10]
```diff
@@ -60,4 +65,33 @@ public static String sanitizingUri(String uri) {
 		}
 		return uri;
 	}
+
+	public static boolean matches(String pattern, String uri, PathMatcher pathMatcher) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/URIUtils.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/URIUtils.java [hunk 11]
```diff
@@ -60,4 +65,33 @@ public static String sanitizingUri(String uri) {
...
+			return matches(pattern, new URI(uri), pathMatcher);
+		} catch (Exception e) {
+			return false;
+		}
+	}
+
+	public static boolean matches(String pattern, URI uri, PathMatcher pathMatcher) {
+		if (pattern.length() < 1) {
+			return false;
+		}
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/URIUtils.java]: remove that

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+	public boolean matches(String uri) {
+		return URIUtils.matches(pattern, uri, pathMatcher);
+	}
+
+	public boolean matches(URI uri) {
+		return URIUtils.matches(pattern, uri, pathMatcher);
+	}
+
+	public static boolean matches(XMLExcludedSymbolFile[] excluded, TextDocument document) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLExcludedSymbolFile.java]: Move this static method to XMLSymbolSettings (no static) isExcluded(String uri) method

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/TextDocument.java [hunk 13]
```diff
@@ -52,6 +54,20 @@ public void setIncremental(boolean incremental) {
 		}
 	}
 
+	/**
+	 * @param doSymbols the doSymbols to set
+	 */
+	public void setDoSymbols(Boolean doSymbols) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/TextDocument.java]: Son't cache it the result of excluded doSymbols. We will do that if we see performance problem but for xml file association the match(uri) is done every time and it works great

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLTextDocumentService.java [hunk 14]
```diff
@@ -412,7 +423,13 @@ public void updateCompletionSettings(CompletionSettings newCompletion) {
 	}
 
 	public void updateSymbolSettings(XMLSymbolSettings newSettings) {
-		sharedSettings.symbolSettings.setEnabled(newSettings.isEnabled());
+		XMLSymbolSettings symbolSettings = sharedSettings.symbolSettings;
+		symbolSettings.setEnabled(newSettings.isEnabled());
+		String[] newPatterns = newSettings.getExcluded();
+		if(newPatterns != null) {
+			symbolSettings.setExcludedFiles(XMLExcludedSymbolFile.getExcludedSymbolFiles(newPatterns));
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLTextDocumentService.java]: It should be just symbolSettings#setExcluded which does the XMLExcludedSymbolFile.getExcludedSymbolFiles(newPatterns)

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLTextDocumentService.java [hunk 15]
```diff
@@ -206,11 +207,21 @@ private XMLFormattingOptions getFormattingSettings(String uri) {
...
+		
+		TextDocument document = getDocument(params.getTextDocument().getUri());
+
+		Boolean doSymbols = document.doSymbols();
+		if(doSymbols == null) {
+			doSymbols = !XMLExcludedSymbolFile.matches(sharedSettings.symbolSettings.getExcludedFiles(), document);
+			document.setDoSymbols(doSymbols);
+		}
+		
+		if(!sharedSettings.symbolSettings.isEnabled() || !doSymbols) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLTextDocumentService.java]: Remove the below code and replace with 

```java
String uri = params.getTextDocument().getUri();
if (!sharedSettings.symbolSettings.isEnabled() || !sharedSettings.symbolSettings.isExcluded(uri))
```

**Comment by NikolasKomonen** [comment]
@angelozerr I've updated the PR with your changes. I created the PathPatternMatcher class and refactored the code.

**Comment by angelozerr** [comment]
> @angelozerr I've updated the PR with your changes. I created the PathPatternMatcher class and refactored the code.

Thanks @NikolasKomonen ! Now the code is perfect. Do you think you could write a test with XMLSymbolSettings#isExcluded method? After that I'm OK to merge it.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLSymbolSettings.java [hunk 16]
```diff
@@ -15,13 +15,52 @@
...
+		XMLExcludedSymbolFile[] exclusions = new XMLExcludedSymbolFile[excluded.length];
+
+		for(int i = 0; i < excluded.length; i++) {
+			exclusions[i] = new XMLExcludedSymbolFile(excluded[i]);
+		}
+
+		excludedFiles = exclusions;
+	}
+
+	public boolean isExcluded(String uri) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLSymbolSettings.java]: Please add javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLSymbolSettings.java [hunk 17]
```diff
@@ -15,13 +15,52 @@
...
+		for (XMLExcludedSymbolFile excludedFile : excludedFiles) {
+			if(excludedFile != null) {
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLSymbolSettings.java]: Why are you doing this test?

**Comment by NikolasKomonen** [comment]
@angelozerr I've updated the pr with tests and javadocs. Could you check it and let me know if I need to add more.

**Comment by angelozerr** [comment]
> @angelozerr I've updated the pr with tests and javadocs. Could you check it and let me know if I need to add more.

It's perfect, thank's @NikolasKomonen !

**Pull request merged**